sunkit_image/time_lag.py

```diff
@@ -0,0 +1,255 @@
+
+    See Also
+    ---------
+    get_lags
+    time_lag
+    max_cross_correlation
+
+    References
+    -----------
+    * https://en.wikipedia.org/wiki/Convolution_theorem
+    * Viall, N.M. and Klimchuk, J.A.
+      Evidence for Widespread Cooling in an Active Region Observed with the SDO Atmospheric Imaging Assembly
+      ApJ, 753, 35, 2012
+      (https://doi.org/10.1088/0004-637X/753/1/35)
+    * Appendix C in Barnes, W.T., Bradshaw, S.J., Viall, N.M.
+      Understanding Heating in Active Region Cores through Machine Learning. I. Numerical Modeling and Predicted Observables
+      ApJ, 880, 56, 2019
+      (https://doi.org/10.3847/1538-4357/ab290c)
+    """
+    # NOTE: it is assumed that the arrays have already been appropriately
+    # interpolated and chunked (if using Dask)
+    delta_lags = np.diff(lags)
+    if not u.allclose(delta_lags, delta_lags[0]):
+        raise ValueError("Lags must be evenly sampled")
+    n_time = (lags.shape[0] + 1) // 2
+    if signal_a.shape != signal_b.shape:
+        raise ValueError("Signals must have same shape.")
+    if signal_a.shape[0] != n_time:
+        raise ValueError("First dimension of signal must be equal in length to time array.")
+    # Reverse the first timeseries
+    signal_a = signal_a[::-1]
+    # Normalize by mean and standard deviation
+    std_a = signal_a.std(axis=0)
+    std_a = np.where(std_a == 0, 1, std_a)  # avoid dividing by zero
+    v_a = (signal_a - signal_a.mean(axis=0)[np.newaxis]) / std_a[np.newaxis]
+    std_b = signal_b.std(axis=0)
+    std_b = np.where(std_b == 0, 1, std_b)
+    v_b = (signal_b - signal_b.mean(axis=0)[np.newaxis]) / std_b[np.newaxis]
+    # Cross-correlation is inverse of product of FFTS (by convolution theorem)
+    fft_a = np.fft.rfft(v_a, axis=0, n=lags.shape[0])
+    fft_b = np.fft.rfft(v_b, axis=0, n=lags.shape[0])
+    cc = np.fft.irfft(fft_a * fft_b, axis=0, n=lags.shape[0])
+    # Normalize by the length of the timeseries
+    return cc / signal_a.shape[0]
+
+
+def _get_bounds_indices(lags, bounds):
+    # The start and stop indices are computed in this way
+    # because Dask does not like "fancy" multidimensional indexing
+    if bounds is not None:
+        (indices,) = np.where(np.logical_and(lags >= bounds[0], lags <= bounds[1]))
+        start = indices[0]
+        stop = indices[-1] + 1
+    else:
+        start = 0
+        stop = lags.shape[0] + 1
+    return start, stop
+
+
+def _check_array(signal, lags):
+    # This is to avoid having to specify time as a Dask array so that it
+    # can be specified as a quantity and so that the time lag can be
+    # returned as a Dask quantity and not computed eagerly
+    try:
+        import dask.array  # do this here so that Dask is not a hard requirement
+    except ImportError:
+        return lags
+    else:
+        if isinstance(signal, dask.array.Array):
+            return dask.array.from_array(lags, chunks=lags.shape)
+        else:
+            return lags
```

Comment:
```suggestion
    try:
        import dask.array
    except ImportError:
        return lags
    if isinstance(signal, dask.array.Array):
        return dask.array.from_array(lags, chunks=lags.shape)
    else:
        return lags
```

I feel like this could be cleaned up further.

Reply:
I'm thinking of throwing this out altogether in favor of users being able to pass in `_pre_check` and `_post_check` hooks to modify the lags array appropriately. This is useful because `time`/`lags` will always be a quantity and doesn't play well with other array types (e.g. Dask, Cupy).